src/riak_kv_app.erl

```diff
@@ -74,40 +103,169 @@ start(_Type, _StartArgs) ->
+                                  {riak_kv,
+                                   legacy_keylisting,
+                                   [{false, false}]}),
+    
+    riak_core_capability:register({riak_kv, listkeys_backpressure},
+                                  [true, false],
+                                  false,
+                                  {riak_kv,
+                                   listkeys_backpressure,
+                                   [{true, true}, {false, false}]}),
+    
+    riak_core_capability:register({riak_kv, index_backpressure},
+                                  [true, false],
+                                  false),
+    
+    %% mapred_system should remain until no nodes still exist
+    %% that would propose 'legacy' as the default choice
+    riak_core_capability:register({riak_kv, mapred_system},
+                                  [pipe],
+                                  pipe,
+                                  {riak_kv,
+                                   mapred_system,
+                                   [{pipe, pipe}]}),
+    
+    riak_core_capability:register({riak_kv, mapred_2i_pipe},
+                                  [true, false],
+                                  false,
+                                  {riak_kv,
+                                   mapred_2i_pipe,
+                                   [{true, true}, {false, false}]}),
+    
+    riak_core_capability:register({riak_kv, anti_entropy},
+                                  [enabled_v1, disabled],
+                                  disabled),
+    
+    riak_core_capability:register({riak_kv, handoff_data_encoding},
+                                  [encode_raw, encode_zlib],
+                                  encode_zlib),
+    
+    riak_core_capability:register({riak_kv, object_format},
+                                  get_object_format_modes(),
+                                  v0),
+    
+    riak_core_capability:register({riak_kv, secondary_index_version},
+                                  [v3, v2, v1],
+                                  v1),
+    
+    riak_core_capability:register({riak_kv, vclock_data_encoding},
+                                  [encode_zlib, encode_raw],
+                                  encode_raw),
+    
+    riak_core_capability:register({riak_kv, crdt},
+                                  [?TOP_LEVEL_TYPES, [pncounter], []],
+                                  []),
+    
+    riak_core_capability:register({riak_kv, crdt_epoch_versions},
+                                  [?E2_DATATYPE_VERSIONS, ?E1_DATATYPE_VERSIONS],
+                                  ?E1_DATATYPE_VERSIONS),
+    
+    riak_core_capability:register({riak_kv, put_fsm_ack_execute},
+                                  [enabled, disabled],
+                                  disabled),
+    
+%    %% register the pipeline capabilities
+    riak_core_capability:register({query_pipeline, create_table},
```

Comment:
Each stage in the pipeline will need to declare what inputs it can receive by means of a capability and what outputs it can emit by checking the downstream capability. I think it will also have to tag what it emits somehow (a record? a tagged tuple) so the downstream can know how to process it...